File: TOP150/Combinations.py

class Solution:

    def combine(self, n: int, k: int) -> List[List[int]]:
        
        choice = [ i for i in range(1 , n+1) ]
        sol = [] 

        def backtrack( choice , track ): 

            if len(track) == k : 
                sol.append(list(track)) 
                return 

            for i , pick in enumerate( choice )  : 

                track.append(pick) 
                backtrack( choice[i+1:]  , track  )
                track.pop()

        backtrack(choice , [])
        return sol 



class Solution:

    def combine(self, n: int, k: int) -> List[List[int]]:
        
        choice = [ i for i in range(1 , n+1) ]
        sol = [] 

        def backtrack( i , j , track ): 

            if len(track) == k : 
                sol.append(list(track)) 
                return 

            for idx , pick in enumerate( choice[i:j] )  : 

                track.append(pick) 
                backtrack(  i+idx+1 ,  j  , track  )
                track.pop()

        backtrack( 0 , n , [])
        return sol 

File: RUSH/Combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */


var combine = function(n, k) {

    let answer = [] ; 

    const backtrack = ( startNum , track ) =>{

        if (track.length === k){
            answer.push( Array.from(track)  ) ; 
            return 
        }

        for (let i = startNum ; i < n+1 ; i++){
            track.push(i) ;
            backtrack( i+1 , track )
            track.pop() ; 
        }
    }
    backtrack( 1 , [] ) ; 
    return answer ; 
};